src/server/stream_family.cc

```diff
@@ -1402,44 +1533,131 @@ std::optional<ReadOpts> ParseReadArgsOrReply(CmdArgList args, ConnectionContext*
     return std::nullopt;
   }
 
+  vector<unordered_map<string_view, StreamIDsItem>> res_sitems(shard_set->size());
+  bool stop = false;
+  auto g_cb = [&](Transaction* t, EngineShard* shard) {
+    if (!opts.read_group) {
+      return OpStatus::OK;
+    }
+    auto sid = shard->shard_id();
+    auto s_args = t->GetShardArgs(sid);
+    unordered_map<string_view, StreamIDsItem> sid_items;
+
+    // get group and consumer
+    for (size_t i = 0; i < s_args.size(); i++) {
+      string_view key = s_args[i];
+
+      streamCG* group = nullptr;
+      streamConsumer* consumer = nullptr;
+      auto group_res = FindGroup(t->GetOpArgs(shard), key, opts.gname);
+      if (!group_res) {
+        (*cntx)->SendError(group_res.status());
+        stop = true;
+        return group_res.status();
+      }
+      if (group = group_res->second; !group) {
+        (*cntx)->SendError(NoGroupError(key, opts.gname));
+        stop = true;
+        return OpStatus::SKIPPED;
+      }
+
+      shard->tmp_str1 = sdscpylen(shard->tmp_str1, opts.cname.data(), opts.cname.size());
+      consumer = streamLookupConsumer(group, shard->tmp_str1, SLC_NO_REFRESH);
+      if (!consumer) {
+        consumer =
+            streamCreateConsumer(group, shard->tmp_str1, NULL, 0, SCC_NO_NOTIFY | SCC_NO_DIRTIFY);
+      }
+      StreamIDsItem sitem;
+      sitem.group = group;
+      sitem.consumer = consumer;
+      auto [_, is_inserted] = sid_items.emplace(key, sitem);
```

Comment:
I would love for C++ to allow `[std::ignore, is_insterted]` in structure bindings but yeah, they didn't like the symmetry with `std::tie`. Ignore my comment, just a small rant.